src/core/optimization_container.jl

```diff
@@ -318,12 +330,25 @@ function build_impl!(container::OptimizationContainer, template, sys::PSY.System
         add_dual_variable!(container, sys, transmission_model)
     end
 
-    for branch_model in values(template.branches)
-        @debug "Building $(get_component_type(branch_model)) with $(get_formulation(branch_model)) formulation"
-        TimerOutputs.@timeit BUILD_PROBLEMS_TIMER "$(get_component_type(branch_model))" begin
-            construct_device!(container, sys, branch_model, transmission_model)
-            @debug get_problem_size(container)
-            add_dual_variable!(container, sys, branch_model)
+    for stage in [ArgumentConstructStage(), ConstraintConstructStage()]
+        for branch_model in values(template.branches)
+            @debug "Building $stage for $(get_component_type(branch_model)) with $(get_formulation(branch_model)) formulation" _group =
+                :ConstructGroup
+            TimerOutputs.@timeit BUILD_PROBLEMS_TIMER "$(get_component_type(branch_model))" begin
+                if validate_available_devices(branch_model, sys)
+                    construct_device!(
+                        container,
+                        sys,
+                        stage,
+                        branch_model,
+                        transmission_model,
+                    )
+                end
+                @debug get_problem_size(container)
+                if stage == ArgumentConstructStage()
+                    add_dual_variable!(container, sys, branch_model)
```

Comment:
This should be moved to the `construct_device!` in the `ConstraintConstructStage` and renamed to `add_constraint_dual` so that it makes more sense to the person reading the code.